Testing our apis (#11)
Ada-Developers-Academy/core-unit-2: testing-our-apis/test-setup.md

# Test Setup

## Introduction

> We can continue to use tests and pytest to test our Flask APIs! Reading, writing, and using pytest for a Flask API isn't very different from using pytest outside a Flask API.

Reviewer: minor rewording to flow a bit better from the Fixture lesson. We've discussed a lot of this functionality in the previous lesson, so it should be less surprising that we'll be using tests in general, and pytest specifically.
```suggestion
We can build on our pytest knowledge to test our Flask APIs! Reading, writing, and using pytest for a Flask API isn't very different from using pytest outside a Flask API.
```

> One main difference is the test setup. In this lesson, we'll configure our Flask projects to use pytest and a test database.

Reviewer: specify that this _is_ the main difference. one main difference leaves the door open for the reader to worry about what other differences are going to pop up.
```suggestion
The main difference is in how we set up our tests. In this lesson, we'll configure our Flask projects to use pytest and a test database.
```

## Hello Books API

This lesson uses the [Hello Books API](https://github.com/AdaGold/hello-books-api).

> 

Reviewer: minor formatting
```suggestion

<br />

```

<details style="max-width: 700px; margin: auto;">
    <summary>
        Before beginning this lesson, the Hello Books API should have the following.
    </summary>

- A `hello_books_development` database
- A `hello_books_test` database

> - A `book` table defined in both databases

Reviewer: We haven't made (and won't explicitly make) the books table in the test database. We will be bypassing alembic's migration tracking and use the create_all/drop_all directly to set up and tear down the database with every run.
```suggestion
- A `book` table defined in the development database
```

- A `Book` model defined

> An `.env` file that contains:

Reviewer: I read `.env` as dotenv, so we can use a instead of an.
```suggestion
A `.env` file that contains:
```

```
SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/hello_books_development
SQLALCHEMY_TEST_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/hello_books_test
```

> Endpoints defined for these RESTful routes. They handle missing books:

Reviewer: same change as in previous lesson
```suggestion
Endpoints defined for these RESTful routes. They can gracefully handle missing books:
```

- `GET` to `/books`
- `POST` to `/books`
- `GET` to `/books/<book_id>`
- `PUT` to `/books/<book_id>`
- `DELETE` to `/books/<book_id>`

The `Book` model and table should have the following columns:

- `id`
- `title`
- `description`

</details>

### !callout-danger

## Read-through First, Repeat Second

We **highly suggest** reading or watching through this lesson fully, before repeating on your own machine.

### !end-callout

### !callout-danger

## Prioritize Familiarity

It may be better to think about this lesson as a resource to be familiar with, rather than commit all new knowledge to memory at once.

### !end-callout

## Navigating Our New `tests` Folder

Create a `tests` folder, and the following files:

- `tests/__init__.py`
- `tests/conftest.py`
- `tests/test_routes.py`

We can do so with the following commands:

```
(venv) $ mkdir tests
(venv) $ touch tests/__init__.py tests/conftest.py tests/test_routes.py
```

> Let's consider what these files will do:

Reviewer: consider sounds more like we have some familiarity and are thinking more deeply now, but we are getting our first look here.
```suggestion
Let's take a look at what these files will do:
```

| <div style="min-width:250px;">File</div> | Responsibility of this file                                                                                                                   |
| ---------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- |

> | `__init__.py`                            | Establishes our `tests` package, so it can be properly connected with the rest of the app folders and files. We will probably leave it empty. |

Reviewer: cut probably
```suggestion
| `__init__.py`                            | Establishes our `tests` package, so it can be properly connected with the rest of the app folders and files. We will leave it empty, as we typically do. |
```

> | `conftest.py`                            | A conventional Flask file that will hold test configurations and common test helper functions                                                 |

Reviewer: conftest.py isn't flask specific. It's a standard file in pytest suites.
```suggestion
| `conftest.py`                            | A standard pytest file that holds test configurations and common test helper functions. Essentially, it is automatically `import`ed before any other test files.                                                  |
```

| `test_routes.py`                         | This file will hold the tests for the code in our `app/routes.py` file.                                                                       |

## `conftest.py`

Let's briefly cover the contents of a `tests/conftest.py` file, and then copy the contents of it into our own project.

```python
import pytest
from app import create_app
from app import db

@pytest.fixture
def app():
    app = create_app({"TESTING": True})

    with app.app_context():

>         yield app

Reviewer: add line to create the database tables (moved from the app init)
```suggestion
        db.create_all()
        yield app
```

    with app.app_context():
        db.drop_all()

@pytest.fixture
def client(app):
    return app.test_client()
```

| <div style="min-width:250px;"> Piece of Code </div> | Notes                                                                                                                                                                                                       |
| --------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `import pytest`                                     | We'll be utilizing pytest fixtures to set up our test app                                                                                                                                                   |
| `from app import ... db`                            | We should import `create_app` and `db` in order to configure those when running the tests                                                                                                                   |

> | `@pytest.fixture`                                   | We'll create and use a pytest fixture named `app`, which will be used in our `client` fixtured (defined below)                                                                                              |

Reviewer: fixtured → fixture
the fixture is defined below the code we are discussing, but defined about the discussion itself. defined later sounds a little more "relative" to me.
```suggestion
| `@pytest.fixture`                                   | We'll create and use a pytest fixture named `app`, which will be used in our `client` fixture (defined later)                                                                                              |
```

| `app = create_app( ... )`                           | When we run our tests, this line will run and create an `app` object. It's using the same `create_app` function defined in our `app/__init__.py` file!                                                      |

> | `{"TESTING": True}`                                 | Here, we're passing in a dictionary to represent a "test config" object. If we check the current implementation of `create_app()` in `app/__init__.py`, it uses this argument only to check if it's truthy. |

Reviewer: don't need the "if we check"
```suggestion
| `{"TESTING": True}`                                 | Here, we're passing in a dictionary to represent a "test config" object. The current implementation of `create_app()` in `app/__init__.py` uses this argument only to check if it's truthy (which this is!). |
```

> | `with app.app_context():`                           | This syntax designates that the following code should have an _application context_, or in this case, have access to objects such as `db`.                                                                  |

Reviewer: Give a bit of a hand-wavey overview of what this does, but still tie it back to the database access that we're doing here.
```suggestion
| `with app.app_context():`                           | This syntax designates that the following code should have an _application context_. This lets various functionality in Flask determine what the current running app is. This is particularly important when accessing the database associated with the app.                                                                   |
```

> | `yield app`                                         | This is pytest fixture syntax which executes the actual test at this moment. `yield app` finishes executing once the test has actually be run.                                                              |

Reviewer: `yield` has now been discussed in the fixtures lesson, so we can be a little more precise in the description
```suggestion
| `yield app`                                         | This fixture suspends here, returning the app for use in tests or other fixtures. The lines after this `yield` statement will run after the test using the app has been completed.                                                              |
```

> | `with app....: db.drop_all()`                       | After the test runs, this code specifies that we should clear our test database from any data that was saved inside our test.                                                                               |

Reviewer: ```suggestion
| `with app....: db.drop_all()`                       | After the test runs, this code specifies that we should clear our test database of any data that was saved inside our test.                                                                               |
```

> | `@pytest.fixture`                                   | We're setting up a second test fixture...                                                                                                                                                                   |

Reviewer: we've seen quite a few fixture declarations now (with the addition of the fixture lesson) so I think we can be more matter-of-fact here
```suggestion
| `@pytest.fixture`                                   | Set up a second test fixture                                                                                                                                                                   |
```

| `def client(app):`                                  | This fixture is named `client`. It will request the existing `app` fixture to run, first.                                                                                                                   |
| `return app.test_client()`                          | The responsibility of this fixture is to make a _test client_, which is an object able to _simulate_ a client making HTTP requests.                                                                         |

### !callout-secondary

## Practical Tip: Reading > Writing

> Understanding how to read this code is more valuable than learning how to write it. Realistically, on many projects, test environments are set up. However, it's valuable to read this code in order to:

Reviewer: same edits as previous lesson with this callout
```suggestion
Understanding how to read this code is more valuable than learning how to write it. On many projects, we set up test environments by referring back to code from a previous project. But we should strive to be able to read this code in order to:
```

- Have a better understanding of how the app runs, which will lead to better debugging
- Get familiar with the process of setting it up, in anticipation of future projects

### !end-callout

### The `client` Fixture

The majority of our tests will use the `client` fixture that was defined in `tests/conftest.py`.

> While we're looking at `tests/conftest.py`, let's see a few ways we'll use `client`:

Reviewer: we've been looking at conftest for a bit, so considering is appropriate, which means we can look at (intentional) the commands rather than just see (automatic) them
```suggestion
While we're considering `tests/conftest.py`, let's look at a few ways we'll use `client`:
```
